text/0000-ptr-to-field.md

```diff
@@ -0,0 +1,210 @@
+struct FieldDescriptor {
+    ...
+}
+
+/// The compiler should prevent user implementations for `Field`,
+/// i.e. only the compiler is allowed to implement `Field`
+/// This is to prevent people from creating fake "fields"
+unsafe trait Field {
+    /// The type that the field is a part of
+    type Parent: ?Sized;
+    /// The type of the field
+    type Type: ?Sized;
+
+    /// The metadata required to get to the field using raw pointers
+    const FIELD_DESCRIPTOR: FieldDescriptor;
+}
+
+trait Project<F: Field> {
+    /// The projected version of Self
+    type Projection;
+
+    fn project(self, field: F) -> Self::Projection;
+}
+
+impl<T: ?Sized> *const T {
+    unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *const F::Type {
+        // make the field pointer, this code is allowed to assume that
+        // self points to a valid instance of T
+        ... 
+    }
+}
+
+impl<T: ?Sized> *mut T {
+    unsafe fn project_unchecked<F: Field<Parent = T>>(self, field: F) -> *mut F::Type {
+        // make the field pointer, this code is allowed to assume that
+        // self points to a valid instance of T
+        ...
+    }
+}
+```
+
+Now we need some syntax to refer to the fields types. Some ideas for the syntax are
+
+* `Parent.field`  // my favorite, as it seems most natural
+* `Parent::field` // bad as it conflicts with associated functions
+* `Parent~field`  // or any other sigil
+
+But syntax discussions are defered until after this RFC has been accepted. This is to focues the RFC's discussion on if this addition is needed and close and lingering soundness holes.
+
+Field types will desugar to a unit type that correctly implements `Field`, like so
+
+```rust
+struct Foo {
+    bar: Bar
+}
+
+struct Foo.bar; // or some other name mangle that doesn't conflict with any other name
+
+impl Field for Foo.bar { ... }
+```
+
+These are the core parts of this proposal. Every other part of this proposal can be postponed or dropped without affecting this feature's core principles.
+
+Using these core parts we can build as a library projections through `Pin<&T>`, `Rc<_>` and more. We can then use this to safely project through smart pointers like so.
+
+```rust
+let foo   : Pin<Box<Foo>    = Box::pin(immovable);
+let foo   : Pin<&mut Foo>   = foo.as_mut();
+let field : Pin<&mut Field> = foo.project(Foo.field);
+```
+But to do safe pin projections we will need to introduce a marker trait. Adding this trait is up for debate.
+```rust
+/// The only people who can implement `PinProjectable` are the creator of the parent type
+/// This allows people to opt-in to allowing their fields to be pin projectable.
+/// The guarantee is that once you create `Pin<P<Parent>>`, all of the same guarantees that
+/// apply to `Pin<P<Parent>>` also apply to `Pin<P<Field>>`
+/// For all `Parent: Unpin`, these can be auto implemented for all of their fields.
+unsafe trait PinProjectable: Field {}
```

Comment:
For a type to opt-into this for one of its fields this needs to be involve that base type.

```
struct Foo {
   bar: Bar,
}

// The type `Foo.bar` is not defined here but somewhere else (in he compiler).
// Thus can't simply implement a trait for it.
// unsafe impl PinProjectable for Foo.bar { }
```

But an alternative formulation could remedy this, I think:

```
unsafe trait PinProjectable<Parent: ?Sized>: Field<Parent=Parent> { }

unsafe impl PinProjectable<Foo> for Foo.bar { }
```

Reply:
Because `Parent` is an associated type of `Field`, we don't need a parameter on `PinProjectable`. If you want to constrain `Parent` via `PinProjectable`, you can do `PinProjectable<Parent = ...>`

We could say that for all intents and purposes, these field types are defined right next to the parent types. This way you can just implement traits for them.

Reply:
Don't trait coherence rules forbid you from manually implementing the trait if the Parent type does not occur explicitely in the trait definition? Afterall, that would be implementing a third-party trait for a third-party type. Or should the *type* of `Foo.bar` be treated as if occurring in the crate where `Foo` is defined?

Reply:
The idea is that `Foo.bar` would *not* be a third part type to the person who defined `Foo`.

So, this exactly
> Or should the type of Foo.bar be treated as if occurring in the crate where Foo is defined?